crates/loro-internal/src/kv_store/mem_store.rs

```diff
@@ -0,0 +1,395 @@
+//! │ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ┘│
+//! └────────────────────────────────────────────────────────────────────────────────────────────┘
+//! ```
+//!
+//! Each Key Value Chunk is encoded as follows:
+//!
+//! ```
+//! ┌───────────────────────────────────────────────────────────────┐
+//! │  Key Value Chunk                                              │
+//! │┌ ─ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ┬ ─ ─ ─ ┐│
+//! │ common prefix len key suffix len│key suffix│value len  value  │
+//! ││       u8        │     u16      │  bytes   │   u16   │ bytes ││
+//! │ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ │
+//! └───────────────────────────────────────────────────────────────┘
+//! ```
+//!
+//! Encoding:
+//! 1. Compress key-value pairs data as Key Value Chunk.
+//! 2. Write offsets for each key-value pair.
+//! 3. Write the number of key-value pairs.
+//! 4. **Compress** the entire block using LZ4.
+//! 5. Calculate and append CRC32 checksum.
+//!
+//! Decoding:
+//! 1. Verify the CRC32 checksum.
+//! 2. **Decompress** the block using LZ4.
+//! 3. Read the number of key-value pairs.
+//! 4. Read offsets for each key-value pair.
+//! 5. Parse individual key-value chunks.
+//!
+//! ### Large Value Block
+//!
+//! Large Value Blocks store a single key-value pair with a large value.
+//!
+//! ```
+//! ┌──────────────────────────┐
+//! │Large Block               │
+//! │┌ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─ ─ │
+//! │  value   Block Checksum ││
+//! ││ bytes │      u32        │
+//! │ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘│
+//! └──────────────────────────┘
+//! ```
+//!
+//! Encoding:
+//! 1. Write the value bytes.
+//! 2. Calculate and append CRC32 checksum.
+//!
+//! Decoding:
+//! 1. Verify the CRC32 checksum.
+//! 2. Read the value bytes.
+//!
+//! We need not encode the length of value, because we can get the whole Block by offset in meta.
+//!
+//! ## Block Meta
+//!
+//! The Block Meta section contains metadata for all blocks in the SSTable.
+//!
+//! ```
+//! ┌────────────────────────────────────────────────────────────┐
+//! │ All Block Meta                                             │
+//! │┌ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─┌ ─ ─ ─┌ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ │
+//! │  block length │ Block Meta │ ...  │ Block Meta │ checksum ││
+//! ││     u32      │   bytes    │      │   bytes    │   u32     │
+//! │ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ┘─ ─ ─ ┘─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ┘│
+//! └────────────────────────────────────────────────────────────┘
+//! ```
+//!
+//! Each Block Meta entry is encoded as follows:
+//!
+//! ```
+//! ┌──────────────────────────────────────────────────────────────────────────────────────┐
+//! │ Block Meta                                                                           │
+//! │┌ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ┐ │
+//! │  block offset │ first key len   first key   is large │ last key len     last key     │
```

Comment:
Can we also trip the common prefix from the last key?

Reply:
yes, it's better

Reply:
Oh, we cannot use the last key, because the block is built dynamically.